### ISSUE TITLE
Insert multiple rows in one querry doesn't seem to work

### ISSUE THREAD

**Issue by adikte** [psycopg/psycopg#275]
OS: Windows 10
Postgres version : 14.2
python version : 3.7
psyopg version : 3.0.11
pip version 22.0.4

I am trying to insert multiple rows into a database a the same time with an insert statement using the execute function.
But it doesn't seem to work, i got either a wrong result or an error.

What i tried : 
1) 
`my_values= (('1', '2'), ('5', '6'))

conn = connect(**self.params)
#add empty table named test with col1 and col2 as VARCHAR
cur = conn.cursor()
cur.execute("CREATE TABLE test (col1 VARCHAR, col2 VARCHAR)")
conn.commit()
cur.close()

#Insert two rows into the table test
cur = conn.cursor()
cur.execute(INSERT INTO test (col1, col2) VALUES (%s, %s), my_values)
conn.commit()
cur.close()
`

In that case the function run with no bug but i get a table with one row where each cell contain a tuple like value 
![test_table](https://user-images.githubusercontent.com/103272268/163367938-169c6d59-7503-49be-8805-2dc3dc04b12d.png)

2)
Just changing the insert statement :

`#Insert two rows into the table test
cur = db.conn.cursor()
cur.execute(INSERT INTO test (col1, col2) VALUES %s, %s, my_values)
db.conn.commit()
cur.close() 
`
In this case i get an error : 
syntax error at or near "$1"
LINE 1: INSERT INTO chose (col1, col2) VALUES $1, $2

**Comment by dvarrazzo** [comment]
In order to insert more than one row, use executemany.

```
cur.executemany("INSERT INTO test (col1, col2) VALUES (%s, %s)", my_values) 
```

**Comment by adikte** [comment]
Thanks you for your answer but is there no way to make it in one SQL statement ? Executemany will make as many insert statements as number of rows to add, which is way slower especially with a large number of rows.

**Comment by dvarrazzo** [comment]
If you want performance you can use [copy](psycopg.org/psycopg3/docs/basic/copy.html). Untested:

```python
with cur.copy("COPY test (col1, col2) FROM STDIN") as copy:
    for record in my_values:
        copy.write_row(record)
```

You can make it a single statement, yes, the syntax is like

```
INSERT INTO test (col1, col2) VALUES (%s, %s), (%s, %s), ... ;
```
You will have to create a string with as many %s groups you need and pass all the values flattened in a single iterable.
COPY is faster and easier to use.

**Comment by adikte** [comment]
Thanks you for your answer. I have been doing a lot of tests but i can't manage to make it work. Most of the types i tried doesn't work, except for string (toand datetime.time for both of the method you gave me.


data type python :  numpy.bool_  column type postgres : BOOLEAN
COPY from stdin failed: error from Python: ProgrammingError - cannot adapt type 'int8' using placeholder '%t' (format: TEXT)

data type python :  numpy.int8_  column type postgres : INTEGER or SMALLINT OR BIGINT
COPY from stdin failed: error from Python: ProgrammingError - cannot adapt type 'bool_' using placeholder '%t' (format: TEXT)

data type python :  pandas._libs.tslibs.timestamps.Timestamp_  column type postgres : DATE
COPY from stdin failed: error from Python: ProgrammingError - cannot adapt type 'datetime64' using placeholder '%t' (format: TEXT)

is their a way around that ?

**Comment by dvarrazzo** [comment]
Numpy types are not supported yet: see #192

**Comment by adikte** [comment]
Thanks and sorry i saw this one  #275 after my post
Both of your method work now with : 

conn.adapters.register_dumper(np.int8, psycopg.types.numeric.IntDumper)
conn.adapters.register_dumper(np.int32, psycopg.types.numeric.IntDumper)
conn.adapters.register_dumper(np.int64, psycopg.types.numeric.IntDumper)
conn.adapters.register_dumper(np.bool_, psycopg.types.bool.BoolDumper)
conn.adapters.register_dumper(np.float16, psycopg.types.numeric.FloatDumper)
conn.adapters.register_dumper(np.float32, psycopg.types.numeric.FloatDumper)
conn.adapters.register_dumper(np.float64, psycopg.types.numeric.FloatDumper)
conn.adapters.register_dumper(np.datetime64,psycopg.types.datetime.DateDumper )

**Comment by dvarrazzo** [comment]
Thank you, that's useful to know.

I haven't played with numpy types and matrices but it would be nice to register these types to dump out of the box. Providing binary dumpers would make it even more efficient.

**Comment by CarMoreno** [comment]
I had the same problem using `pd.Int64DType` which is used to manage an integer column with missing values. I resolve it using the comment let by @adikte 

`connection.adapters.register_dumper(pd.Int64Dtype, psycopg.types.numeric.IntDumper)`

But, for some reason psycopg3 has problem to identify `pd.NA` values as missing values (`<NA>`). Therefore I had to convert these values to `None` using `pd.replace({pd.NA: None}, inplace=True)`.

**Comment by dvarrazzo** [comment]
What is the MRO of this NA object?

**Comment by jamesdow21** [comment]
@dvarrazzo Quite a bit late to answer this question, but I found this issue when looking for any details on supporting pd.NA

pd.NA is a special singleton of the type pandas._libs.missing.NAType (full MRO is (pandas._libs.missing.NAType, pandas._libs.missing.C_NAType, object))

It is used to represent the missing values for any Series with a nullable datatype  
Since pandas 1.5+ supports Series backed by pyarrow arrays (in which all data types are nullable) instead of only numpy arrays (where only float and datetime were nullable)

I would imagine that pd.NA should be special cased to always correspond to postgresql NULL (similar to how python None is handled) 
That's at least how I'm trying to handle it (but running into some difficulties)

**Comment by dvarrazzo** [comment]
@jamesdow21 at the moment it's not possible for a dumper to dump a NULL, see #377.

This was being addressed in #642, on which @canhuynh1998 was working. I don't know it that MR is stuck, but we want to release none-returning dumpers in 3.2.

**Comment by jamesdow21** [comment]
Glad to hear! Allowing a dumper to return None would work really well (especially if there's a variety of known None-like singletons)

In the meantime, I actually took your advice from earlier in the thread and converted both of my current use cases to COPY queries

The PostgreSQL default null '\N' bytes works in a `COPY ... FROM STDIN` query

For anyone else stumbling upon this later:  
```python
import pandas as pd
import psycopg
from pandas._libs.missing import NAType
from psycopg.sql Import Identifier, SQL

class NADumper(Dumper):
    def dump(self, obj: NAType) -> bytes:
        return b"\\N"

psycopg.adapters.register_dumper(NAType, NADumper)

...

df = pd.DataFrame(...)

columns = SQL(", ").join(map(Identifier, df.columns))
copy_query = SQL("copy table_name ({}) from stdin;").format(columns)

with psycopg.connect(...) as conn:
    with conn.cursor() as cur:
        with cur.copy(copy_query) as copy:
            for row in df.itertuples(index=False, name=None):
                copy.write_row(row)
```
